src/nvim/assert.h

```diff
@@ -121,4 +121,33 @@
     ((enum { ASSERT_CONCAT(assert_line_, __LINE__) = 1/(!!(e)) }) 0)
 #endif
 
+/// @def STRICT_ADD
+///
+/// Requires GCC 5+ and Clang 3.8+
+///   https://clang.llvm.org/docs/LanguageExtensions.html
+///   https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html
+///
+/// Alternative for compilers without __builtin_xx_overflow ?
+///   https://stackoverflow.com/a/44830670/152142
+///
+/// @param MAX Maximum value of the narrowest type of operand.
+///            Not used if compiler supports __builtin_add_overflow.
+#if (defined(__clang__) && __has_builtin(__builtin_add_overflow)) \
+  || (__GNUC__ >= 5)
+# define STRICT_ADD(a, b, c, t) \
+  do { if (__builtin_add_overflow(a, b, c)) { abort(); } } while (0)
+#else
+# define STRICT_ADD(a, b, c, t) \
+  do { *c = (t)(a + b); } while (0)
```

Comment:
Why not `*c = (t)(a) + (b)`, which is more inline with PVS' suggestion.

Reply:
@jamessan Then we run into `-Wsign-conversion` and `-Wconversion` errors. 

So we satisfy PVS and improve correctness on newer systems, but older systems will just need to live with the old approach.